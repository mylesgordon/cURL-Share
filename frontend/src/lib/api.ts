import { backendUrl } from './stores';
import { get } from 'svelte/store';
import type { Project, ProjectInfo } from './types';

const url = get(backendUrl);

type CreateProjectResponse = {
	id: number;
};

type ProjectAdminStatus = {
	isUserAdmin: boolean;
}

type Fetch = typeof fetch;

export async function fetchIsLoggedIn(fetch: Fetch): Promise<boolean> {
	const response = await fetch(`${url}/api/v1/user-status`, {
		method: 'GET',
		mode: 'cors',
		headers: { 'Access-Control-Allow-Origin': url },
		credentials: 'include'
	});
	const responseJson = await response.json();
	return responseJson.is_logged_in ? responseJson.is_logged_in : false;
}

export async function fetchProject(fetch: Fetch, projectId: string): Promise<Project> {
	const projectResponse = await fetch(`${url}/api/v1/project/${projectId}`, {
		method: 'GET',
		mode: 'cors',
		headers: { 'Access-Control-Allow-Origin': url },
		credentials: 'include'
	});

	return await projectResponse.json();
}

export async function fetchProjectAdminStatus(fetch: Fetch, projectId: string): Promise<ProjectAdminStatus> {
	const projectResponse = await fetch(`${url}/api/v1/project/${projectId}/is-user-admin`, {
		method: 'GET',
		mode: 'cors',
		headers: { 'Access-Control-Allow-Origin': url },
		credentials: 'include'
	});

	return await projectResponse.json();
}

export async function fetchProjects(fetch: Fetch): Promise<ProjectInfo[]> {
	const response = await fetch(`${url}/api/v1/project`, {
		method: 'GET',
		mode: 'cors',
		headers: { 'Access-Control-Allow-Origin': url },
		credentials: 'include'
	});
	return await response.json();
}

export async function createProjectRequest(
	fetch: Fetch,
	projectInfo: ProjectInfo
): Promise<CreateProjectResponse> {
	const response = await fetch(`${url}/api/v1/project`, {
		method: 'POST',
		body: JSON.stringify(projectInfo),
		mode: 'cors',
		headers: { 'Access-Control-Allow-Origin': url, 'Content-Type': 'application/json' },
		credentials: 'include'
	});

	return await response.json();
}

export async function deleteProjectRequest(fetch: Fetch, project_id: number): Promise<void> {
	await fetch(`${url}/api/v1/project/${project_id}`, {
		method: 'DELETE',
		mode: 'cors',
		headers: { 'Access-Control-Allow-Origin': url },
		credentials: 'include'
	});
}

export async function updateProjectRequest(fetch: Fetch, project: Project): Promise<void> {
	await fetch(`${url}/api/v1/project/${project.info.id}`, {
		method: 'POST',
		body: JSON.stringify(project),
		mode: 'cors',
		headers: { 'Access-Control-Allow-Origin': url, 'Content-Type': 'application/json' },
		credentials: 'include'
	});
}

export async function logInRequest(
	fetch: Fetch,
	endpoint: string,
	username: string,
	password: string
): Promise<number> {
	const request = await fetch(`${url}/api/v1/${endpoint}`, {
		method: 'POST',
		mode: 'cors',
		headers: {
			'Access-Control-Allow-Origin': url,
			'Content-Type': 'application/json'
		},
		body: JSON.stringify({ username, password }),
		credentials: 'include'
	});

	return request.status;
}

export async function logOutRequest(fetch: Fetch): Promise<void> {
	await fetch(`${url}/api/v1/log-out`, {
		method: 'POST',
		mode: 'cors',
		headers: { 'Access-Control-Allow-Origin': url },
		credentials: 'include'
	});
}
