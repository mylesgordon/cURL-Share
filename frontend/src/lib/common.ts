export function splitAndTrim(input: string): Array<string> {
		return input.split(',').map((item) => item.trim());
	}