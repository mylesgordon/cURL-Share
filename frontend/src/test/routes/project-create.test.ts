import { createProjectRequest } from '$lib/api.page';
import { getByText, render } from '@testing-library/svelte';
import { goto } from '$app/navigation';
import { screen } from '@testing-library/svelte';
import { testMatchingSnapshot } from '../common';
import ProjectCreatePage from '../../routes/project/create/+page.svelte';
import userEvent from '@testing-library/user-event';
import type { CreateProjectResponse, ProjectInfo } from '$lib/types';

vi.mock('$app/navigation');
vi.mock('$lib/api.page');

describe('Project create page', () => {
	function getPageElements() {
		const nameField = screen.getByLabelText('Name');
		const descriptionField = screen.getByLabelText('Description');
		const createButton = screen.getByRole('button', { name: /Create/i });

		return { nameField, descriptionField, createButton };
	}

	testMatchingSnapshot(ProjectCreatePage);

	it('should display the correct title', () => {
		render(ProjectCreatePage);
		expect(document.title).toEqual('Create Project | cURL Share');
	});

	it('should display an error message when creating a project with no name or description', async () => {
		render(ProjectCreatePage);
		const { nameField, descriptionField, createButton } = getPageElements();

		const table = [
			{ fieldWithText: nameField, fieldToClear: descriptionField },
			{ fieldWithText: descriptionField, fieldToClear: nameField }
		];

		for (const { fieldWithText, fieldToClear } of table) {
			await userEvent.type(fieldWithText, 'cool text');
			await userEvent.clear(fieldToClear);
			await userEvent.click(createButton);

			await screen.getByText('Name or description fields empty - please try again.');
		}
	});

	// TODO
	// it('should display an error message if the server sends a bad response',async () => {
	// 	render(ProjectCreatePage);
	// 	const {nameField, descriptionField, createButton} = getPageElements();
	// 	const error = new Error("Something went wrong!");
	// 	createProjectRequest.mockRejectedValue(error);

	// 	await userEvent.type(nameField, "my project");
	// 	await userEvent.type(descriptionField, "this is a jolly project");
	// 	await userEvent.click(createButton);

	// 	await screen.findByText("Could not create project - please try again.");
	// });

	it('should send the correct data and redirect if project creation is successful', async () => {
		const requestObject: ProjectInfo = {
			id: 0,
			environments: '',
			description: 'Cool project',
			name: 'Project 1',
			visibility: 'Private'
		};
		const mockResolvedResponse: CreateProjectResponse = { id: 1 };

		render(ProjectCreatePage);
		const { nameField, descriptionField, createButton } = getPageElements();
		const privateButton = await screen.getByText('Private');

		createProjectRequest.mockResolvedValue(mockResolvedResponse);

		await userEvent.type(nameField, requestObject.name);
		await userEvent.type(descriptionField, requestObject.description);
		await userEvent.click(privateButton);
		await userEvent.click(createButton);

		expect(createProjectRequest).toHaveBeenCalledWith(window.fetch, requestObject);
		expect(goto).toHaveBeenCalledWith('/project/1');
	});
});
