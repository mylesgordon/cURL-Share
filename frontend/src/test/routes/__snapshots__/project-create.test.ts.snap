// Vitest Snapshot v1

exports[`Project create page > should match snapshot 1`] = `
<body>
  <div>
    <!--&lt;Meta&gt;-->
     
    <h2>
      Create a new project
    </h2>
     
    <form
      class="flex flex-col space-y-3 pt-4"
    >
      <div
        class="w-100 s-JZthjkXlZrq3"
      >
        <label
          class="label s-JZthjkXlZrq3"
          for="name"
        >
          Name
        </label>
         
        <input
          class="input input-rounded s-JZthjkXlZrq3"
          id="name"
          type="text"
        />
         
      </div>
      <!--&lt;Input&gt;-->
       
      <div
        class="w-100 s-JZthjkXlZrq3"
      >
        <label
          class="label s-JZthjkXlZrq3"
          for="description"
        >
          Description
        </label>
         
        <textarea
          class="input input-rounded s-JZthjkXlZrq3"
          id="description"
        />
         
      </div>
      <!--&lt;Input&gt;-->
       
      <fieldset
        class="radio-group s-K6jzjY3vjtlg"
      >
        <legend
          class="radio-legend s-K6jzjY3vjtlg"
        >
          Visibility
        </legend>
         
        <label
          class="radio-label-wrap s-K6jzjY3vjtlg"
        >
          <input
            class="radio s-K6jzjY3vjtlg"
            id="visibility-visibility-0"
            name="visibility"
            type="radio"
            value="Public"
          />
           
          <span
            aria-hidden="true"
            class="radio-label s-K6jzjY3vjtlg"
          />
           
          <span
            class="radio-label-copy s-K6jzjY3vjtlg"
          >
            Public
          </span>
           
        </label>
        <label
          class="radio-label-wrap s-K6jzjY3vjtlg"
        >
          <input
            class="radio s-K6jzjY3vjtlg"
            id="visibility-visibility-1"
            name="visibility"
            type="radio"
            value="Private"
          />
           
          <span
            aria-hidden="true"
            class="radio-label s-K6jzjY3vjtlg"
          />
           
          <span
            class="radio-label-copy s-K6jzjY3vjtlg"
          >
            Private
          </span>
           
        </label>
      </fieldset>
      <!--&lt;ChoiceInput&gt;-->
      <!--&lt;VisibilityInput&gt;-->
       
       
      <button
        class="btn btn-primary btn-bordered btn-rounded s-NAQuwh5ehqbV"
        type="submit"
      >
        Create
      </button>
      
      <!--&lt;Button&gt;-->
    </form>
    <!--&lt;+page&gt;-->
  </div>
</body>
`;
